### ISSUE TITLE
Decode to a premultiplied form

### ISSUE THREAD

**Issue by jrmuizel** [image-rs/image-png#271]
Consumers of PNG images often want the alpha premultiplied. See https://github.com/jrmuizel/raqote/issues/167

**Comment by fintelia** [comment]
If you want premultiplied alpha, isn't the right approach to apply it yourself before saving the PNG?

**Comment by jrmuizel** [comment]
No, PNGs are stored using unpremultiplied alpha. Premultiplying before saving will cause other png decoders to [apply the alpha twice](https://searchfox.org/mozilla-central/source/image/SurfaceFilters.h#71) and won't help with existing images.

**Comment by jrmuizel** [comment]
I've clarified the title to make it clearer that this would probably be an optional output transformation.

**Comment by HeroicKatora** [comment]
There are at least two case where applying the transformation at decode time, instead of later, is advantageous compared to a post-processing step after the full buffer is decoded. There is no general benefit however, so their implementation mustn't have too much of a complexity overhead.

* For images with a palette it can be performed on the palette which is possibly faster.
* For images with `tRNS` chunks the alpha bit is a single bit. Applying a masking SIMD-AND might be faster than the equivalent full multiplications.

**Comment by jrmuizel** [comment]
Applying at decode time can also avoid cache thrashing on large images.

**Comment by yanchith** [comment]
Sorry for the noob question: If `png` pre-multiplies alpha when decoding, how should this interact with gamma/sRGB? Should the pre-multiply happen in linear space? If so, would `png` have to gamma-decode, pre-multiply, and then gamma-encode again?
